store/postgres/migrations/2019-03-11-221621_delete_unneeded_entity_history/up.sql

```diff
@@ -0,0 +1,190 @@
+-- either data_before or data_after can be null, and we need to make sure
+-- we include rows with a null in either column, but that 'anything !=
+-- null' will always be false. We lift SQL NULL to JSON null, since JSON
+-- uses normal, not three-state logic for comparisons with null
+alter table eh set unlogged;
+insert into eh
+  select id, event_id, entity_id, subgraph, entity, data_before,
+         reversion, op_id from entity_history
+  where subgraph != 'subgraphs'
+    and coalesce(data_before, 'null'::jsonb)
+        != coalesce(data_after, 'null'::jsonb);
+alter table eh set logged;
+
+-- Swap old and new tables
+alter table entity_history rename to eh_old;
+alter table eh rename to entity_history;
+
+-- Create indexes
+alter index entity_history_pkey rename to eh_old_pkey;
+alter table entity_history
+  add constraint entity_history_pkey primary key(id);
+
+alter index entity_history_event_id_btree_idx
+  rename to eh_old_event_id_btree_idx;
+create index entity_history_event_id_btree_idx
+  on entity_history(event_id);
+
+--
+-- Deal with event_meta_data
+--
+
+-- Remove uneeded event_meta_data entries following the same pattern
+create table emd(like event_meta_data
+                     including defaults
+                     including comments
+                     including storage);
+alter table event_meta_data
+      alter column id drop default;
+alter sequence event_meta_data_id_seq
+      owned by emd.id;
+
+-- Copy the good event_meta_data
+alter table emd set unlogged;
+insert into emd
+  select * from event_meta_data d
+  where exists (select 1 from entity_history h where d.id = h.event_id);
+alter table emd set logged;
+
+alter table event_meta_data rename to emd_old;
+alter table emd rename to event_meta_data;
+
+-- Create indexes
+alter index event_meta_data_pkey rename to emd_old_pkey;
+alter table event_meta_data
+  add constraint event_meta_data_pkey primary key(id);
+
+alter index event_meta_data_db_transaction_id_key
+  rename to emd_old_db_transaction_id_key;
+alter table event_meta_data
+  add constraint event_meta_data_db_transaction_id_key
+                 unique(db_transaction_id);
+
+-- Make sure the fk constraint on entity_history points to the new
+-- event_meta_data table
+alter table eh_old
+  drop constraint entity_history_event_id_fkey;
+
+alter table entity_history
+  add constraint entity_history_event_id_fkey
+  foreign key (event_id) references event_meta_data(id)
+  match full on update cascade on delete cascade;
+
+-- Recreate the entity_history_with_source view so that
+-- it points to the right version of these tables
+drop view entity_history_with_source;
+create view entity_history_with_source as
+  select
+    h.subgraph,
+    h.entity,
+    h.entity_id,
+    h.data_before,
+    h.op_id,
+    m.source
+  from entity_history h, event_meta_data m
+  where h.event_id = m.id
+  order by h.event_id desc;
+
+-- Clean up
+drop table eh_old;
+drop table emd_old;
+
+analyze entity_history;
```

Comment:
What does this do?

Reply:
It updates Postgres' internal statistics about the table (important for the query planner) Since we deleted so much data, these statistics will be all wrong. The execution of the statement itself is pretty fast (~ 100ms)